# STA130 Week 01 Homework 

Please see the course [wiki-textbook](https://github.com/pointOfive/stat130chat130/wiki) for the list of topics covered in this homework assignment, and a list of topics that might appear during ChatBot conversations which are "out of scope" for the purposes of this homework assignment (and hence can be safely ignored if encountered)


<details class="details-example"><summary style="color:blue"><u>Introduction</u></summary>

### Introduction
    
A reasonable characterization of STA130 Homework is that it simply defines a weekly reading comprehension assignment. 
Indeed, STA130 Homework essentially boils down to completing various understanding confirmation exercises oriented around coding and writing tasks.
However, rather than reading a textbook, STA130 Homework is based on ChatBots so students can interactively follow up to clarify questions or confusion that they may still have regarding learning objective assignments.

> Communication is a fundamental skill underlying statistics and data science, so STA130 Homework based on ChatBots helps practice effective two-way communication as part of a "realistic" dialogue activity supporting underlying conceptual understanding building. 

It will likely become increasingly tempting to rely on ChatBots to "do the work for you". But when you find yourself frustrated with a ChatBots inability to give you the results you're looking for, this is a "hint" that you've become overreliant on the ChatBots. Your objective should not be to have ChatBots "do the work for you", but to use ChatBots to help you build your understanding so you can efficiently leverage ChatBots (and other resources) to help you work more efficiently.<br><br>

</details>

<details class="details-example"><summary style="color:blue"><u>Instructions</u></summary>

### Instructions
    
1. Code and write all your answers (for both the "Prelecture" and "Postlecture" HW) in a python notebook (in code and markdown cells) 
    
> It is *suggested but not mandatory* that you complete the "Prelecture" HW prior to the Monday LEC since (a) all HW is due at the same time; but, (b) completing some of the HW early will mean better readiness for LEC and less of a "procrastentation cruch" towards the end of the week...
    
2. Paste summaries of your ChatBot sessions (including link(s) to chat log histories if you're using ChatGPT) within your notebook
    
> Create summaries of your ChatBot sessions by using concluding prompts such as "Please provide a summary of our exchanges here so I can submit them as a record of our interactions as part of a homework assignment" or, "Please provide me with the final working verson of the code that we created together"
    
3. Save your python jupyter notebook in your own account and "repo" on [github.com](github.com) and submit a link to that notebook though Quercus for assignment marking<br><br>

</details>

<details class="details-example"><summary style="color:blue"><u>Prompt Engineering?</u></summary>

### Prompt Engineering?    
    
The questions (as copy-pasted prompts) are designed to initialize appropriate ChatBot conversations which can be explored in the manner of an interactive and dynamic textbook; but, it is nonetheless **strongly recommendated** that your rephrase the questions in a way that you find natural to ensure a clear understanding of the question. Given sensible prompts the represent a question well, the two primary challenges observed to arise from ChatBots are 

1. conversations going beyond the intended scope of the material addressed by the question; and, 
2. unrecoverable confusion as a result of sequential layers logial inquiry that cannot be resolved. 

In the case of the former (1), adding constraints specifying the limits of considerations of interest tends to be helpful; whereas, the latter (2) is often the result of initial prompting that leads to poor developments in navigating the material, which are likely just best resolve by a "hard reset" with a new initial approach to prompting.  Indeed, this is exactly the behavior [hardcoded into copilot](https://answers.microsoft.com/en-us/bing/forum/all/is-this-even-normal/0b6dcab3-7d6c-4373-8efe-d74158af3c00)...

</details>


### Marking Rubric (which may award partial credit) 

- [0.1 points]: All relevant ChatBot summaries [including link(s) to chat log histories if you're using ChatGPT] are reported within the notebook
- [0.2 points]: Reasonable well-written general definitions for Question "2.2"
- [0.3 points]: Demonstrated understanding regarding Question "4"
<!-- - [0.2 points]: A sensible justification for the choice in Question "7.4" -->
- [0.4 points]: Requested assessment of ChatBot versus google performance in Question "8.3"


## "Prelecture" HW [*completion prior to next LEC is suggested but not mandatory*]

#### 1. Pick one of the datasets from the ChatBot session(s) of the **TUT demo** (or from your own ChatBot session if you wish) and use the code produced through the ChatBot interactions to import the data and confirm that the dataset has missing values<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> If your TA has not shared a relevant ChatBot session from their **TUT demo** through a piazza post and a Quercus announcement, the **TUT notebook** has links to example ChatBot sessions that you can use; or, ...
> 
> ```python
> # feel free to just use the following if you prefer...
> import pandas as pd
> url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
> df = pd.read_csv(url)
> df.isna().sum()
> ```
    
</details>

#### 2. Start a new ChatBot session with an initial prompt introducing the dataset you're using and request help to determine how many columns and rows of data a `pandas` DataFrame has, and then

1. use code provided in your ChatBot session to print out the number of rows and columns of the dataset; and,  
2. write your own general definitions of the meaning of "observations" and "variables" based on asking the ChatBot to explain these terms in the context of your dataset<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> A good initial prompt to start would be be something like
> - "I've downloaded a dataset about characters from animal crossings (from https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv), and I'd like to know what columns of information I have and how much data I have"
> 
> You can further reduce the scope of your inquiry with if needed with something like
> - "I've already downloaded the data and want to understand the size (or dimensions) of the dataset to start with"
> 
> *Some ChatBots can upload your data and do this for you; but, extended usage of this feature [likely requires a paid subscription](https://github.com/pointOfive/stat130chat130/blob/main/CHATLOG/wk1/GPT/SLS/00006_gpt3p5_LoadDataPaywall.md); and, anyway, you need to run the code yourself rather than having a ChatBot do that for you; and, for STA130 we don't want a ChatBot to just do the analysis for us; rather, we instead want ChatBots to help us understand the steps we need to take to analyze the data; so,* **you DO NOT need to purchase an upgraded version of any ChatBots**
> - Free-tier level ChatBots like [GPT4o-mini](https://chat.openai.com/) or [Copilot](https://copilot.microsoft.com/) (which is partially based on [ChatGPT4.0](https://chat.openai.com/), and which you have access to through your UofT account) are sufficiently sophisticated and perfectly appropriate for the STA130 course
    
</details>

#### 3. Ask the ChatBot how you can provide simple summaries of the columns in the dataset and use the suggested code to provide these summaries for your dataset<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> Use your ChatBot session to help you create working examples of using  `df.describe()` and `df['column'].value_counts()` for your dataset (although note that the `.value_counts()` method is not really meant to be used for numeric variables, so if you dataset has only numeric variables, `.value_counts()` might not be particularly informative...)
>
> #### ChatBot Response Scope 
>     
> If prompts are not sufficiently focused you will likely get overly broad responses from the ChatBot, but you can always respond with subsequent refinement requests to appropriately limit the scope of the ChatBot responses to focus on addressing your actual content targets; so, 
> - an initially very general inquiry like, "I need help analyzing my data" will likely result in a ChatBot response suggesting a wide variety of approaches and techniques for summarizing your dataset; but, re-prompting the ChatBot with something like, "What's the simplest form of summarization of this dataset that I could do and how do I do it in Python?" or suggesting guidance using the specific summarization methods requested above will helpfully re-orient the ChatBot to your specific interests and needs
> 
> #### Jupyter Notebook Hints
> 
> Jupyter notebook printouts usaully don't show all of the data (when there's too much to show, like if `df.describe()` includes results for many columns), but the printouts just show enough of the data to give an idea of what the results are which is all we're looking for at the moment
> 
> - Consider dividing the code that ChatBot provides you into different jupyter notebook cells so that each cell concludes with a key printed result; the last line of code in a jupyter notebook cell will automatically print out in a formatted manner, so replacing something like `print(df.head())` with `df.head()` as the last line of a cell provides a sensible way to organize your code
> - The printout suggestions above are demonstrated in `STA130F24_CourseProject.ipynb` if looking at an example would be helpful to understand what they're getting at...
    
</details>
                
#### 4. If the dataset you're using has (a) non-numeric variables and (b) missing values in numeric variables, explain (perhaps using help from a ChatBot if needed) the discrepancies between size of the dataset given by `df.shape` and what is reported by `df.describe()` with respect to (a) the number of columns it analyzes and (b) the values it reports in the "count" column<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> If the dataset you're using does not have (a) non-numeric variables and (b) missing values in numeric variables (e.g., the `"villagers.csv"` example above has only a single numeric variable `row_n` which has no missing values), instead download and use the [https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv](https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv)" data to answer this question  
>
> In (a) above, the "columns it analyzes" refers to the columns of the output of `df.describe()` which will only include "numeric" columns by default, but you can can see the names of all the columns in a dataset using `df.columns`; and, make sure `df.shape` is refering to the dataset you think it is... if you've loaded a different dataset it might not have been called `df`(!)
>
> **If you get any errors (for example related to column names), copy and paste them as a response to the ChatBot, and see if it can help you resove them by adding the suggested adjustments to your code and then reruning all your code to see if the changes have fixed the problem (and repeat this process as needed until the problems have been resolved).**
    
</details>
    
#### 5. Use your ChatBot session to help understand the difference between the following and then provide your own paraphrasing summarization of that difference

- an "attribute", such as `df.shape` which does not end with `()`
- and a "method", such as `df.describe()` which does end with `()` 
   

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> The fact that a "method" such as `df.describe()` ends with `()` suggests that "methods" are essentially something that we would call a "function" in programming language terminology; but, without getting too technical or "in the weeds", it might also be worth considering that we could also contrast what the difference is between a "function" in a programming language versus a "function" in mathematics...  
    
</details><br><br>

***Don't forget to ask for summaries of your ChatBot session(s) and paste these into your homework notebook (including link(s) to chat log histories if you're using ChatGPT)!***<br><br>

In [6]:
#part 2.1, code provided from ChatGPT

import pandas as pd

# Load the dataset
file_path = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv'  # Replace this with your dataset's path
df = pd.read_csv(file_path)

# Show basic information about the dataset
print("### Basic Information ###")
print(df.info())  # Summary of columns, non-null values, and data types

# Display the first few rows
print("\n### First 5 Rows ###")
print(df.head())  # Displays the first 5 rows

# Get the number of rows and columns
rows, cols = df.shape
print(f"\nThe dataset contains {rows} rows and {cols} columns.")

# Display column names
print("\n### Column Names ###")
print(df.columns)

# Summary statistics for numerical columns
print("\n### Summary Statistics ###")
print(df.describe())

### Basic Information ###
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 391 entries, 0 to 390
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   row_n        391 non-null    int64 
 1   id           390 non-null    object
 2   name         391 non-null    object
 3   gender       391 non-null    object
 4   species      391 non-null    object
 5   birthday     391 non-null    object
 6   personality  391 non-null    object
 7   song         380 non-null    object
 8   phrase       391 non-null    object
 9   full_id      391 non-null    object
 10  url          391 non-null    object
dtypes: int64(1), object(10)
memory usage: 33.7+ KB
None

### First 5 Rows ###
   row_n       id     name  gender    species birthday personality  \
0      2  admiral  Admiral    male       bird     1-27      cranky   
1      3  agent-s  Agent S  female   squirrel      7-2       peppy   
2      4    agnes    Agnes  female        pig  

part 2.2
An observation in the context of data sets are data points, organized in rows. Each observation is a single data point, for example in the animal crossings data set, each observation is one villager.
Variables, on the other hand, are measured attributes of the observations, i.e. the data collected about the observations. They are the columns in a data set.

In [7]:
#part 3

import pandas as pd

# Load the dataset
file_path = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'  # Replace with your dataset's path
df = pd.read_csv(file_path)

# Get the number of rows and columns
rows, cols = df.shape
print(f"\nThe dataset contains {rows} rows and {cols} columns.")\

# 1. Basic Information
print("### Basic Information ###")
print(df.info())  # Gives an overview of the dataset

# 2. First few rows
print("\n### First 5 Rows ###")
print(df.head())  # Shows the first 5 rows

# 3. Shape of the dataset
print(f"\nThe dataset contains {df.shape[0]} rows and {df.shape[1]} columns.")

# 4. Column Names
print("\n### Column Names ###")
print(df.columns)

# 5. Summary statistics for numerical variables
print("\n### Summary Statistics (Numerical Variables) ###")
print(df.describe())  # Displays summary statistics like mean, std, min, etc., for numerical columns

# 6. Summary statistics for categorical variables
print("\n### Summary Statistics (Categorical Variables) ###")
categorical_columns = df.select_dtypes(include=['object']).columns
print(df[categorical_columns].describe())  # Displays stats like unique values, top, frequency for categorical columns

# 7. Check for missing values
print("\n### Missing Values ###")
print(df.isnull().sum())  # Shows the count of missing values for each column

# 8. Value counts for categorical variables
print("\n### Value Counts for Categorical Variables ###")
for column in categorical_columns:
    print(f"\nValue counts for {column}:")
    print(df[column].value_counts())

# 10. Data types of columns
print("\n### Data Types of Columns ###")
print(df.dtypes)  # Displays data types of all columns



The dataset contains 891 rows and 15 columns.
### Basic Information ###
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   sex          891 non-null    object 
 3   age          714 non-null    float64
 4   sibsp        891 non-null    int64  
 5   parch        891 non-null    int64  
 6   fare         891 non-null    float64
 7   embarked     889 non-null    object 
 8   class        891 non-null    object 
 9   who          891 non-null    object 
 10  adult_male   891 non-null    bool   
 11  deck         203 non-null    object 
 12  embark_town  889 non-null    object 
 13  alive        891 non-null    object 
 14  alone        891 non-null    bool   
dtypes: bool(2), float64(2), int64(4), object(7)
memory usage: 92.4+ KB
None

### First 5 Rows ###
   surviv

part 4.a
There's a discrepancy between df.describe() and df.shape. df.describe() only contains 6 columns, while df.shape contains 15 columns. 
Since only numerical columns are included in df.describe, other columns like "who", "embark_town", "alive", etc, are left out,
causing many more columns in df.shape than df.describe().

4.b
In addition, df.shape says there should be 891 rows, meaning 891 values in the "count", one for each observation. However, the "age" column from df.describe()
only has 714 values in "count".
In other words, the data set is incomplete and missing some ages for some passengers aboard, and so df.describe does not include it in its "count", literally.

part 5

An attribute is a property of an object that stores a piece of data, and accessing it does not analyze the data in any way.
In this way, an attribute simply provides access to already existing stored information.
an attribute does not require parentheses.

A method is a function of an object. Unlike attributes, methods perform some sort of action on the object when called.
Such as df.describe(), which must calculate the mean from the data instead of simply accessing it.
Methods require parentheses, even if there is no arguments passed to it.

<details class="details-example"><summary style="color:blue"><u>Continue now...?</u></summary>

### Prelecture VS Postlecture HW
    
Feel free to work on the "Postlecture" HW below if you're making good progress and want to continue: in this case this is particularly reasonable as questions "6" and "7" below directly follow up and extend the "Prelecture" HW questions

*The benefits of continue would are that (a) it might be fun to try to tackle the challenge of working through some problems without additional preparation or guidance; and (b) this is a very valable skill to be comfortable with; and (c) it will let you build experience interacting with ChatBots (and beginning to understand their strengths and limitations in this regard)... it's good to have sense of when using a ChatBot is the best way to figure something out, or if another approach (such as course provided resources or a plain old websearch for the right resourse) would be more effective*
    
</details>    


## "Postlecture" HW [*submission along with "Prelecture" HW is due prior to next TUT*]

#### 6. The `df.describe()` method provides the 'count', 'mean', 'std', 'min', '25%', '50%', '75%', and 'max' summary statistics for each variable it analyzes. Give the definitions (perhaps using help from the ChatBot if needed) of each of these summary statistics<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> The answers here actually make it obvious why these can only be calculated for numeric variables in a dataset, which should help explain the answer to "4(a)" and "4(b)" above
>   
> Also notice that when `df.describe()` is used missing values are not explicitly removed, but `df.describe()`  provides answers anyway. Is it clear what `df.describe()` does with the data in each columns it analyzes if there is missing data in the column in question? 
>
> The next questions addresses removing rows or columns from a dataset in order to explicitly remove the presense of any missingness in the dataset (assuming we're not going to fill in any missing data values using any missing data imputation methods, which are beyond the scope of STA130); so, the behavior of `df.describe()` hints that explicitly removing missing may not always be necessary; but, the concern, though, is that not all methods may be able to handle missing data the way `df.describe()` does...
    
</details>

#### 7. Missing data can be considered "across rows" or "down columns".  Consider how `df.dropna()` or `del df['col']` should be applied to most efficiently use the available non-missing data in your dataset and briefly answer the following questions in your own words

1. Provide an example of a "use case" in which using `df.dropna()` might be peferred over using `del df['col']`<br><br>
    
2. Provide an example of "the opposite use case" in which using `del df['col']` might be preferred over using `df.dropna()` <br><br>
    
3. Discuss why applying `del df['col']` before `df.dropna()` when both are used together could be important<br><br>
    
4. Remove all missing data from one of the datasets you're considering using some combination of `del df['col']` and/or `df.dropna()` and give a justification for your approach, including a "before and after" report of the results of your approach for your dataset.<br><br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> Start a new ChatBot session **[but remember to first ask your ChatBot for summaries of your current session and perhaps coding results (so you can supply these in the homework as requested)]**, since your last ChatBot session has likely gotten quite long and has covered a lot of material at this point 
> - It can sometimes be helpful to reset ChatBot sessions to refocus them on the topics of inquiry without too much backlog history that might unintentionally bias things in certain directions and, of course, you can always re-introduce material from earlier conversations as it's relevant, such as for answering "D" based on reintroducing and updating code you made in a previous ChatBot session.  
> 
> #### ChatBot Scope Guidance
> 
> - This question is not interested in the general benefits of imputing missing data, or the general benefits of using `df.dropna()` and/or `del df['col']` to remove missing data, just how to most efficiently remove missing data if a user chooses to do so
> 
> - More sophisticated analyses for "filling in" rather than removing missing data (as considered here) are possible (based on making assumptions about missing data and using specific imputation methods or models) but these are "beyond the scope" of this homework assignment so this topics can be safely ignored for now
> 
> #### ChatBot Code Troubleshooting
> 
> A key issue to be aware of when asking ChatBots for help with something is that they are not running and checking code for correctess, and they often intertwine written instructions with code instructions; so, BEFORE YOU RUN ANY CODE provided by a ChatBot, you should check the following
> 
> 1. If this code changes an object or data, are you sure you want to run this code?
> 2. Can you easily "undo" the results of running code (e.g., from a copy `df_saved=df.copy()` or reloading the data) if running the code doesn't do what you want?
> 3. Is the state of the data what is expected by the code? Or have the objects been updated and changed so they're no longer what the code expects them to be? 
> 
> #### If you get any `Python` errors, copy and paste them into the ChatBot prompt and see if it can help you resove them; but, keep in mind the final point above becasue the ChatBot might not be aware of the state of your objects relative to the code it's producing...

</details><br>
    
#### 8. Give brief explanations in your own words for any requested answers to the questions below

> This problem will guide you through exploring how to use a ChatBot to troubleshoot code using the "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv" data set 
> 
> To initialially constrain the scope of the reponses from your ChatBot, start a new ChatBot session with the following slight variation on the initial prompting approach from "2" above
> - "I am going to do some initial simple summary analyses on the titanic data set I've downloaded (https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv) which has some missing values, and I'd like to get your help understanding the code I'm using and the analysis it's performing"
        
1. Use your ChatBot session to understand what `df.groupby("col1")["col2"].describe()` does and then demonstrate and explain this using a different example from the "titanic" data set other than what the ChatBot automatically provide for you
    
> If needed, you can help guide the ChatBot by showing it the code you've used to download the data **AND provide it with the names of the columns** using either a summary of the data with `df.describe()` or just `df.columns` as demonstrated [here](../CHATLOG/COP/00017_copilot_groupby.md)
    
2. Assuming you've not yet removed missing values in the manner of question "7" above, `df.describe()` would have different values in the `count` value for different data columns depending on the missingness present in the original data.  Why do these capture something fundamentally different from the values in the `count` that result from doing something like `df.groupby("col1")["col2"].describe()`?

> Questions "4" and "6" above address how missing values are handled by `df.describe()` (which is reflected in the `count` output of this method); but, `count` in conjunction with `group_by` has another primary function that's more important than addressing missing values (although missing data could still play a role here).

3. Intentionally introduce the following errors into your code and report your opinion as to whether it's easier to (a) work in a ChatBot session to fix the errors, or (b) use google to search for and fix errors: first share the errors you get in the ChatBot session and see if you can work with ChatBot to troubleshoot and fix the coding errors, and then see if you think a google search for the error provides the necessary toubleshooting help more quickly than ChatGPT<br><br>
    
    1. Forget to include `import pandas as pd` in your code 
       <br> 
       Use Kernel->Restart from the notebook menu to restart the jupyter notebook session unload imported libraries and start over so you can create this error
       <br><br>
       When python has an error, it sometimes provides a lot of "stack trace" output, but that's not usually very important for troubleshooting. For this problem for example, all you need to share with ChatGPT or search on google is `"NameError: name 'pd' is not defined"`<br><br>

    2. Mistype "titanic.csv" as "titanics.csv"
       <br> 
       If ChatBot troubleshooting is based on downloading the file, just replace the whole url with "titanics.csv" and try to troubleshoot the subsequent `FileNotFoundError: [Errno 2] No such file or directory: 'titanics.csv'` (assuming the file is indeed not present)
       <br><br>
       Explore introducing typos into a couple other parts of the url and note the slightly different errors this produces<br><br>
      
    3. Try to use a dataframe before it's been assigned into the variable
       <br> 
       You can simulate this by just misnaming the variable. For example, if you should write `df.groupby("col1")["col2"].describe()` based on how you loaded the data, then instead write `DF.groupby("col1")["col2"].describe()`
       <br><br>
       Make sure you've fixed your file name so that's not the error any more<br><br>
        
    4. Forget one of the parentheses somewhere the code
       <br>
       For example, if the code should be `pd.read_csv(url)` the change it to `pd.read_csv(url`<br><br>
        
    5. Mistype one of the names of the chained functions with the code 
       <br>
       For example, try something like `df.group_by("col1")["col2"].describe()` and `df.groupby("col1")["col2"].describle()`<br><br>
        
    6. Use a column name that's not in your data for the `groupby` and column selection 
       <br>
       For example, try capitalizing the columns for example replacing "sex" with "Sex" in `titanic_df.groupby("sex")["age"].describe()`, and then instead introducing the same error of "age"<br><br>
        
    7. Forget to put the column name as a string in quotes for the `groupby` and column selection, and see if the ChatBot and google are still as helpful as they were for the previous question
       <br>
       For example, something like `titanic_df.groupby(sex)["age"].describe()`, and then `titanic_df.groupby("sex")[age].describe()`
        
#### 9. Have you reviewed the course [wiki-textbook](https://github.com/pointOfive/stat130chat130/wiki) and interacted with a ChatBot (or, if that wasn't sufficient, real people in the course piazza discussion board or TA office hours) to help you understand all the material in the tutorial and lecture that you didn't quite follow when you first saw it?<br>
    
<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> Just answering "Yes" or "No" or "Somewhat" or "Mostly" or whatever here is fine as this question isn't a part of the rubric; but, the midterm and final exams may ask questions that are based on the tutorial and lecture materials; and, your own skills will be limited by your familiarity with these materials (which will determine your ability to actually do actual things effectively with these skills... like the course project...)
    
</details>
    
***Don't forget to ask for summaries of your ChatBot session(s) and paste these into your homework notebook (including link(s) to chat log histories if you're using ChatGPT)!***

# Q6


For each variable (column) within a dataframe, DataFrame.describe() provides:

Count - The count summary statistics simply describes how many non-null values are within a column, 
i.e. how many data points there are

Mean - This is the arithmetic mean, the sum of all data points divided by the number of data points, i.e. the average.

Std - The standard deviation, which describes the spread of the data, how far apart they are from the mean.
Higher standard deviation means further apart from the mean.

Min - The minimum value in a column.

25% - The first quartile of the data. Only 25% of the values fall within below the first quartile, which can help 
understand the spread of lower values.

50% - The second quartile of the data, also known as the median value. 50% of the data points are below this, and on the
hand 50% of the values are above

75% - The third quartile of the data, of which 75% of the values are below. Can help understand the spread of the upper values

Max - The maximum value of a column.

## Q7


7.1 df.dropna() simply removes all rows with missing values, while del df['col'] deletes columns.
df.dropna() could discard problematic observations with missing values
instead of entirely discarding possibly important variables. For example, if your variable has a sample of students
and only some students are missing their age, it would be much more useful to drop those students instead of the entire 'age' column.

7.2 On the other hand, del df['col'] discards entire columns, which still has its use cases. If a data set has entirely
useless or unnecessary variables, or missing data points are concentrated in one column, it would be pragmatic to remove the column itself.
For example in a data set containing shoe sales in Canada. If trying to determine the average shoe size of a Canadian, it may be
practical to remove the price of the shoe when analyzing the data set.


7.3 Applying del drop['col'] before df.dropna() is important mainly because it reduces data loss. If a column has many missing values
and df.dropna() is applied first, you could lose a substantial amount of data points, even if the column would've been deleted the step after.
If del drop['col'] is utilized first, those data points missing values in 'col' are preserved.


In [8]:
#7.4, using data set from https://www.kaggle.com/datasets/NUFORC/ufo-sightings/data (UFO sightings)

import pandas as pd

url = 'https://github.com/heklly/sta130/raw/main/UFOdata.csv'
df = pd.read_csv(url, on_bad_lines='skip', low_memory=False)

##Before removing empty data
print(df.shape)
print(df.isna().sum())
df

(88679, 11)
datetime                    0
city                        0
state                    7409
country                 12365
shape                    2922
duration (seconds)          2
duration (hours/min)     3019
comments                   35
date posted                 0
latitude                    0
longitude                   0
dtype: int64


datetime                  city state country     shape  \
0      10/10/1949 20:30            san marcos    tx      us  cylinder   
1      10/10/1949 21:00          lackland afb    tx     NaN     light   
2      10/10/1955 17:00  chester (uk/england)   NaN      gb    circle   
3      10/10/1956 21:00                  edna    tx      us    circle   
4      10/10/1960 20:00               kaneohe    hi      us     light   
...                 ...                   ...   ...     ...       ...   
88674    9/9/2013 22:00                  napa    ca      us     other   
88675    9/9/2013 22:20                vienna    va      us    circle   
88676    9/9/2013 23:00                edmond    ok      us     cigar   
88677    9/9/2013 23:00                 starr    sc      us   diamond   
88678    9/9/2013 23:30        ft. lauderdale    fl      us      oval   

      duration (seconds) duration (hours/min)  \
0                   2700           45 minutes   
1                   7200              1-2 hrs   
2                     20           20 seconds   
3                     20             1/2 hour   
4                    900           15 minutes   
...                  ...                  ...   
88674               1200                 hour   
88675                  5            5 seconds   
88676               1020           17 minutes   
88677                  0             2 nights   
88678                  0       still occuring   

                                                comments date posted  \
0      This event took place in early fall around 194...   4/27/2004   
1      1949 Lackland AFB&#44 TX.  Lights racing acros...  12/16/2005   
2      Green/Orange circular disc over Chester&#44 En...   1/21/2008   
3      My older brother and twin sister were leaving ...   1/17/2004   
4      AS a Marine 1st Lt. flying an FJ4B fighter/att...   1/22/2004   
...                                                  ...         ...   
88674                                       Napa UFO&#44   9/30/2013   
88675  Saw a five gold lit cicular craft moving fastl...   9/30/2013   
88676  2 witnesses 2  miles apart&#44 Red &amp; White...   9/30/2013   
88677  On September ninth my wife and i noticed stran...   9/30/2013   
88678  Hovering object lit with red and white lights&...   9/30/2013   

         latitude   longitude  
0      29.8830556  -97.941111  
1        29.38421  -98.581082  
2            53.2   -2.916667  
3      28.9783333  -96.645833  
4      21.4180556 -157.803611  
...           ...         ...  
88674  38.2972222 -122.284444  
88675  38.9011111  -77.265556  
88676  35.6527778  -97.477778  
88677  34.3769444  -82.695833  
88678  26.1219444  -80.143611  

[88679 rows x 11 columns]

In [9]:
#7.4
#From the data above, not counting observations with multiple missing variables, only about 26k/89k data points are missing
#With such a large data set, I've decided to simply drop the points with missing data. 63k observations is more than
#enough for any analysis, I felt. Also, information like country/state, despite having lots of missing data was too important
#to simply drop in my opinion.
import pandas as pd

url = 'https://github.com/heklly/sta130/raw/main/UFOdata.csv'
df = pd.read_csv(url, on_bad_lines='skip', low_memory=False)
scrubbed_df = df.dropna()

#After removing empty data
print(scrubbed_df.shape)
print(scrubbed_df.isna().sum())
scrubbed_df

(69528, 11)
datetime                0
city                    0
state                   0
country                 0
shape                   0
duration (seconds)      0
duration (hours/min)    0
comments                0
date posted             0
latitude                0
longitude               0
dtype: int64


datetime            city state country     shape  \
0      10/10/1949 20:30      san marcos    tx      us  cylinder   
3      10/10/1956 21:00            edna    tx      us    circle   
4      10/10/1960 20:00         kaneohe    hi      us     light   
5      10/10/1961 19:00         bristol    tn      us    sphere   
7      10/10/1965 23:45         norwalk    ct      us      disk   
...                 ...             ...   ...     ...       ...   
88674    9/9/2013 22:00            napa    ca      us     other   
88675    9/9/2013 22:20          vienna    va      us    circle   
88676    9/9/2013 23:00          edmond    ok      us     cigar   
88677    9/9/2013 23:00           starr    sc      us   diamond   
88678    9/9/2013 23:30  ft. lauderdale    fl      us      oval   

      duration (seconds) duration (hours/min)  \
0                   2700           45 minutes   
3                     20             1/2 hour   
4                    900           15 minutes   
5                    300            5 minutes   
7                   1200           20 minutes   
...                  ...                  ...   
88674               1200                 hour   
88675                  5            5 seconds   
88676               1020           17 minutes   
88677                  0             2 nights   
88678                  0       still occuring   

                                                comments date posted  \
0      This event took place in early fall around 194...   4/27/2004   
3      My older brother and twin sister were leaving ...   1/17/2004   
4      AS a Marine 1st Lt. flying an FJ4B fighter/att...   1/22/2004   
5      My father is now 89 my brother 52 the girl wit...   4/27/2007   
7      A bright orange color changing to reddish colo...   10/2/1999   
...                                                  ...         ...   
88674                                       Napa UFO&#44   9/30/2013   
88675  Saw a five gold lit cicular craft moving fastl...   9/30/2013   
88676  2 witnesses 2  miles apart&#44 Red &amp; White...   9/30/2013   
88677  On September ninth my wife and i noticed stran...   9/30/2013   
88678  Hovering object lit with red and white lights&...   9/30/2013   

         latitude   longitude  
0      29.8830556  -97.941111  
3      28.9783333  -96.645833  
4      21.4180556 -157.803611  
5      36.5950000  -82.188889  
7      41.1175000  -73.408333  
...           ...         ...  
88674  38.2972222 -122.284444  
88675  38.9011111  -77.265556  
88676  35.6527778  -97.477778  
88677  34.3769444  -82.695833  
88678  26.1219444  -80.143611  

[69528 rows x 11 columns]

# Q8

In [10]:
#8.1

url = 'https://github.com/heklly/sta130/raw/main/UFOdata.csv'
df = pd.read_csv(url, on_bad_lines='skip', low_memory=False)
scrubbed_df = df.dropna()
scrubbed_df.groupby('country')['state'].describe()


count unique top  freq
country                        
au          10      7  nt     2
ca        3096     15  on  1383
gb          11     10  ns     2
us       66411     52  ca  9064

Groupby combines data points that are the same value within a column ("col1") and groups them together into a singular group/row for all unique values in col1. Groupby then indexes values of 'col 2' to make new columns in relation to that row, creating an entirely new object. Then .describe() works on that object to provide summary statistics.

In this case, I decided to group the country UFO sightings are seen, then by the state.
Count gives me the number of sightings within each country, and unique tells me how many unique states UFO sightings were seen.
Top gives me the most frequent state within each country UFO sightings are seen, and freq is the number seen in that state.

In [11]:
#8.2
import pandas as pd

url = 'https://github.com/heklly/sta130/raw/main/UFOdata.csv'
df = pd.read_csv(url, on_bad_lines='skip', low_memory=False)
print(df.shape)
print(df[['country', 'state']].describe())
df.groupby('country')['state'].describe()

(88679, 11)
       country  state
count    76314  81270
unique       5     68
top         us     ca
freq     70293  10449


count unique  top  freq
country                         
au          11      7   nt     3
ca        3250     15   on  1443
de           0      0  NaN   NaN
gb          12     11   ns     2
us       70293     52   ca  9575

#8.2 

df.describe()'s count value counts all non-empty data points in a column. From which, you can tell what number of data
points are missing from a variable. However, df.groupby("col1")["col2"].describe() is fundamentally different in that
the 'count' provides the number of the data points with the same value within a column ("col1"). In other words, counting the
frequency of some value in a column rather than providing the number of non-empty data points.

In [12]:
#8.3
import pandas as pd

url = 'https://github.com/heklly/sta130/raw/main/UFOdata.csv'
df = pd.read_csv(url, on_bad_lines='skip', low_memory=False)
print(df.shape)
print(df[[country, 'state']].describe())
df.groupby('country')['state'].describe()

(88679, 11)


NameError: name 'country' is not defined

## 8.3
A. ChatGPT is faster, but Google also gives me the solution pretty quickly

B. Again, using ChatGPT is just easier and  more efficient. Google's results, while is helpful gives results not specific to my code

C. ChatGPT is faster and tells me exactly what is wrong, but Google is also pretty fast.

D. ChatGPT can fix my code correctly, but I can't even begin to Google my issue this time

E. ChatGPT can fix my issue, Google also has the correct solution because it seems to be an error someone else came across

F. ChatGPT tells me to recheck my column names, and so does Google. Both equally useful

G. ChatGPT recognizes that the name needs to be in quotes, but Google is not very useful

## 9. Yes, I've used ChatGPT to help me understand some of the material

# Recommended Additional Useful Activities [Optional]

The "Ethical Profesionalism Considerations" and "Current Course Project Capability Level" sections below **are not a part of the required homework assignment**; rather, they are regular weekly guides covering (a) relevant considerations regarding professional and ethical conduct, and (b) the analysis steps for the STA130 course project that are feasible at the current stage of the course<br><br>

<details class="details-example"><summary style="color:blue"><u>Ethical Professionalism Considerations</u></summary>

## Ethical Professionalism Considerations

> If the observed data is "no events occured" does this mean the data is "missing" and [should be ignored](https://priceonomics.com/the-space-shuttle-challenger-explosion-and-the-o)?
> 
> - NASA: \<determines temperature doesn't affects "o-ring" by subseting data to just "o-ring" incidents\>
> - Also NASA: \<launches the shuttle on a cold day\>

|No apparent "o-ring" failure and temperature relationship|Apparent between "o-ring" failure and temperature relationship|
|:-|:-|
if you just look at "o-ring" failure event data|if you instead look at ALL the data as you should|
|![](https://etzq49yfnmd.exactdn.com/wp-content/uploads/2022/03/image06-14.png)|![](https://etzq49yfnmd.exactdn.com/wp-content/uploads/2022/03/image02-33.png)|
|![](https://upload.wikimedia.org/wikipedia/commons/8/8b/Shuttle_Challenger_explosion.gif?20190203170223)|![](https://i.makeagif.com/media/10-04-2014/nT57xW.gif)|

<br>
    
</details>    

<details class="details-example"><summary style="color:blue"><u>Current Course Project Capability Level</u></summary>

## Current Course Project Capability Level

> The data we'll use for the STA130 course project is based on the [Canadian Social Connection Survey](https://casch.org/cscs). Please see the [data use agreement](https://static1.squarespace.com/static/60283c2e174c122f8ebe0f39/t/6239c284d610f76fed5a2e69/1647952517436/Data+Use+Agreement+for+the+Canadian+Social+Connection+Survey.pdf) regarding the appropriate and ethical professional use of this data (available at the bottom of the [CSCS](https://casch.org/cscs) webpage).
> 
> 1. Have a very quick look at the list of available variables available using the [link](https://drive.google.com/file/d/1ISVymGn-WR1lcRs4psIym2N3or5onNBi/view) (again at the bottom of the [CSCS](https://casch.org/cscs) webpage); then, 
> 2. examine the code in the first thirteen code cells of [STA130F24_CourseProject.ipynb](https://github.com/pointOfive/stat130chat130/blob/main/CP/STA130F24_CourseProject.ipynb) to get an initital understanding of how we might subset to different studies included in the [data](https://drive.google.com/file/d/1mbUQlMTrNYA7Ly5eImVRBn16Ehy9Lggo/view) (again accessible at the bottom of the [CSCS](https://casch.org/cscs) webpage); then,     
> 3. review the fourteenth and fifteenth cells (with the comments "Here's a high level summary of the data" and "And here are some explanations about the columns in the data") a little more closely to get a better sense of which columns seem to be the most interesting and whether or not they seem to have a lot of missing data
    
</details>        

### Afterward

Here are few ideas of some other kinds of interactions you might consider exploring with a ChatBot...

> While these are likely to be extremely practically valuable, they are not a part of the homework assignment, so do not include anything related to these in your homework submission

- With respect to improving ones ability in statistics, coding, communication, and other key data science skills
    - what is the ChatBots perception its own capabilities and uses as an AI-driven assistance tool 
    - and does ChatBots assessment of itself influence or agree with your own evalution of the ChatBot? 

- ChatBots can introduce and explain the "World War 2 planes" problem and the "Monte Hall" problem... 
    - how well does do they seem to do and introducing and explaining other "unintuitive surprising statistics paradoxes"?

- If you consider the process of writing about why you chose to take this course, and the skills you were hoping to build through this course with respect to your current ideas about what possible careers 
    - and how do you think the exercise would be different if you framed it as a dialogue with a ChatBot
    - and do you think the difference could be positive and productive, or potentially biasing and distracting?
    
- ChatBots sometimes immediately responds in simple helpful ways, but other times it gives a lot of extraneous information that can be overwheling... are you able to prompt and interact with ChatBots in manner that keeps its reponses helpful and focused on what you're interested in? 

- ChatBots tends to respond in a fairly empathetic and supportive tone...
    - do you find it helpful to discuss concerns you might have about succeeding in the course (or entering university more generally) with a ChatBot?
    
- For what purposes and in what contexts do you think a ChatBot could provide suggestions or feedback about your experiences that might be useful? 
